drivers: add ATWINC15x0 WiFi netdev driver

gschorcht opened the pull request:
### Contribution description

This PR adds a `netdev` driver for the Microchip ATWINC15x0 WiFi module.

**Background**

The ATWINC15x0 WiFi module is used on several boards and extension boarsd, for example:

- [Adafruit Feather M0 WiFi](https://learn.adafruit.com/adafruit-feather-m0-wifi-atwinc1500/)
- [ATWIN1500-XPRO](https://www.microchip.com/DevelopmentTools/ProductDetails/ATWINC1500-XPRO) for the extension of Atmel Xplained PRO boards
- [Adafruit ATWINC1500 WiFi breakout](https://learn.adafruit.com/adafruit-atwinc1500-wifi-module-breakout)
- [Watterott ATWINC1500-Breakout](https://github.com/watterott/ATWINC1500-Breakout)

Furthermore, the ATWINC15x0 WiFi module is the only module supported by the standard [Arduino WiFi101 library](https://github.com/arduino-libraries/WiFi101). Therefore, there are a number of Arduino shields with the ATWINC15x0 WiFi module, e.g. [Adafruit WINC1500 WiFi Shield for Arduino](https://learn.adafruit.com/adafruit-winc1500-wifi-shield-for-arduino).

Using this `netdev` driver together with an ATWINC15x0 WiFi module enables the connection via WiFi in RIOT. Thus, ATWINC15x0 WiFi module can be used as `netdev` for GNRC as well as lwIP `netif`s.

**Current Status**

The driver works out of the box for the following boards that have the ATWINC15x0 on-board:
  - [Adafruit Feather M0 WiFi](https://doc.riot-os.org/group__boards__feather-m0.html)
  - [Arduino MKR1000](https://doc.riot-os.org/group__boards__arduino-mkr1000.html)

It was also tested with an [Adafruit WINC1500 WiFi Shield for Arduino](https://learn.adafruit.com/adafruit-winc1500-wifi-shield-for-arduino) for ESP32 and STM32 Nucleo-F411RE.

Arduino AVR boards unfortunately have too less RAM.

The driver supports following WiFi security modes:
- WPA2 Personal mode (PSK)
- Open System.

WPA2 enterprise mode with IEEE 802.1X/EAP authentication is not supported for the following reasons:
- At least ATWINC15x0 firmware version 19.6.1 is required, but most available modules still have a firmware version lower than 19.5.0. Older firmware versions support EAP-TTLSv0/MSCHAPv2 but use an invalid SSL/TLS version.
- At least version 19.6.1 of the ATWINC15x0 vendor driver is required. The driver version used by this netdev driver, which is available as open source, is version 19.5.2.

**Future Work**

At the moment, only station mode is implemented. It might be interesting to support AP mode so that it could be used in a border router.

### Testing procedure

It should work with `examples/gnrc_networking`
```
USEMODULE='atwinc15x0' \
CFLAGS='-DWIFI_SSID=\"ssid\" -DWIFI_PASS=\"pass\" \
        -DATWINC15X0_PARAM_SSN_PIN=GPIO_PIN\(1,6\) \
        -DATWINC15X0_PARAM_RESET_PIN=GPIO_PIN\(1,4\) \
        -DATWINC15X0_PARAM_IRQ_PIN=GPIO_PIN\(0,8\)' \
make BOARD=... -C examples/gnrc_networking flash term
```

### Issues/PRs references

aabadie commented:
Is it a take-over from #7433 ?

Otherwise thanks for providing this. I'm be able to test this PR on arduino-mkr1000.

gschorcht commented:
> Is it a take-over from #7433 ?

No, I didn't know that. Even though I searched for all combinations `winc`, `atwinc`, `atwinc1500`, `winc1500`, `atwinc15x0`, `winc15x0`, ... in issues and PRs to avoid duplicates, I didn't find it. Probably because I didn't search in closed issues and PRs.

> I'm be able to test this PR on arduino-mkr1000.

Sounds good.

aabadie commented:
> Even though I searched for all combinations winc, atwinc, atwinc1500, winc1500, atwinc15x0, winc15x0, ... in issues and PRs to avoid duplicates, I didn't find it. Probably because I didn't search in closed issues and PRs.

You can blame the stale bot ;)

gschorcht commented:
> > I'm be able to test this PR on arduino-mkr1000.

@aabadie Thanks for pointing out this board. I found such a board, which I bought 3 years ago, unused in a box. I will try the driver with this board to refine the default parameters of the driver a little bit.

gschorcht commented:
> You can blame the stale bot ;)

But at that time there was also no reaction from the author for over a year. So closing the PR was correct.

aabadie commented:
> But at that time there was also no reaction from the author for over a year. So closing the PR was correct.

I know, it was a joke :)

gschorcht commented:
@aabadie My first test with mkr1000 did not work, I have to check whether I used the right GPIOs.

gschorcht commented:
@aabadie Good news. Worked for me for mkr1000 with following configuration. I had to use `SPI_DEV(1)` instead of `SPI_DEV(0)`.
```
USEMODULE='atwinc15x0' \
CFLAGS='-DGNRC_IPV6_NIB_CONF_SLAAC=1 -DDEBUG_ASSERT_VERBOSE \
        -DWIFI_SSID=\"<ssid>\" -DWIFI_PASS=\"<pass>\" \
        -DATWINC15X0_PARAM_SPI=SPI_DEV\(1\) \
        -DATWINC15X0_PARAM_SSN_PIN=GPIO_PIN\(0,14\) \
        -DATWINC15X0_PARAM_RESET_PIN=GPIO_PIN\(0,27\) 
        -DATWINC15X0_PARAM_IRQ_PIN=GPIO_PIN\(1,9\) \
        -DATWINC15X0_PARAM_CHIP_EN_PIN=GPIO_PIN\(0,28\) \
        -DATWINC15X0_PARAM_WAKE_PIN=GPIO_PIN\(1,8\)' \
make BOARD=arduino-mkr1000 -C examples/gnrc_networking flash
```
```
> ifconfig
Iface  7  HWaddr: F8:F0:05:F5:E1:50  Link: up 
L2-PDU:1500 MTU:1440  HL:255  RTR  
RTR_ADV  
Source address length: 6
Link type: wireless
inet6 addr: fe80::faf0:5ff:fef5:e150  scope: link  VAL
inet6 addr: 2003:c1:e703:555:faf0:5ff:fef5:e150  scope: global  VAL
inet6 group: ff02::2
inet6 group: ff02::1
inet6 group: ff02::1:fff5:e150

          Statistics for Layer 2
          Statistics for IPv6
```
I don't know why the output of `ifconfig' is not correctly indented. On ESP32 the indentations are correct.

Anyway, the driver work with mkr1000 and we could enable WiFi in board definition once the driver has been finished and is merged.

aabadie commented:
Well done @gschorcht!

> I don't know why the output of `ifconfig' is not correctly indented

I think this problem comes from the stdio over USB layer. For some reasons, the `\t` the character is dropped.

benpicco commented:
> @aabadie Good news. Worked for me for mkr1000 with following configuration
> 
> ```C
> #define ATWINC15X0_PARAM_SPI         SPI_DEV(1)
> #define ATWINC15X0_PARAM_SSN_PIN     GPIO_PIN(0,14) 
> #define ATWINC15X0_PARAM_RESET_PIN   GPIO_PIN(0,27) 
> #define ATWINC15X0_PARAM_IRQ_PIN     GPIO_PIN(1,9)
> #define ATWINC15X0_PARAM_CHIP_EN_PIN GPIO_PIN(0,28)
> #define ATWINC15X0_PARAM_WAKE_PIN    GPIO_PIN(1,8)
> ```

better add that to the board's `board.h` then

gschorcht commented:
> I think this problem comes from the stdio over USB layer. For some reasons, the `\t` character is dropped.

It seem that it donsn't like spaces directly after the newline. The following function is responsible for the line break and the indentation:
https://github.com/RIOT-OS/RIOT/blob/e738508308068e7d324ecbdaa41485aff28937cd/sys/shell/commands/sc_gnrc_netif.c#L364-L371

gschorcht commented:
> > ```c
> > #define ATWINC15X0_PARAM_SPI         SPI_DEV(1)
> > #define ATWINC15X0_PARAM_SSN_PIN     GPIO_PIN(0,14) 
> > #define ATWINC15X0_PARAM_RESET_PIN   GPIO_PIN(0,27) 
> > #define ATWINC15X0_PARAM_IRQ_PIN     GPIO_PIN(1,9)
> > #define ATWINC15X0_PARAM_CHIP_EN_PIN GPIO_PIN(0,28)
> > #define ATWINC15X0_PARAM_WAKE_PIN    GPIO_PIN(1,8)
> > ```
> 
> better add that to the board's `board.h` then

That was my intention, of course. I just wanted to try out the settings on the command line first. I also suppose the ATWINC15x0 `netdev` should be defined `netdev_default` device.

gschorcht commented:
> > ```c
> > #define ATWINC15X0_PARAM_SPI         SPI_DEV(1)
> > #define ATWINC15X0_PARAM_SSN_PIN     GPIO_PIN(0,14) 
> > #define ATWINC15X0_PARAM_RESET_PIN   GPIO_PIN(0,27) 
> > #define ATWINC15X0_PARAM_IRQ_PIN     GPIO_PIN(1,9)
> > #define ATWINC15X0_PARAM_CHIP_EN_PIN GPIO_PIN(0,28)
> > #define ATWINC15X0_PARAM_WAKE_PIN    GPIO_PIN(1,8)
> > ```
> better add that to the board's `board.h` then

And I will do the same for the Adafruit feather M0 WiFi once I have tested.

@benpicco BTW, now it's time to introduce `WIFI_SSID` and `WIFI_PATH` as common configurations :wink:

gschorcht commented:
> > I think this problem comes from the stdio over USB layer. For some reasons, the `\t` character is dropped.
> 
> It seem that it donsn't like spaces directly after the newline. The following function is responsible for the line break and the indentation:
> https://github.com/RIOT-OS/RIOT/blob/e738508308068e7d324ecbdaa41485aff28937cd/sys/shell/commands/sc_gnrc_netif.c#L364-L371

```diff
-        printf("\n          ");
+        puts("");
+        printf("          ");
```
solves the problem even if it doesn't seem to elegant. Should I provide it as a fix?

gschorcht commented:
@aabadie With the last commit, mkr1000 should work out of the box with
```
CFLAGS='-DWIFI_SSID=\"<ssid>\" -DWIFI_PASS=\"<pass>\"' make BOARD=arduino-mkr1000 -C examples/gnrc_networking
```

aabadie commented:
And also the documentation is very well written :)

gschorcht commented:
> Just tested on arduino-mkr1000 on my personal WiFi (with IPv6 on) and it works:
> 
> But I don't understand why the board doesn't automatically configures a global IP only link local is setup. Any idea @gschorcht ?

It worked for me. Sometimes it can take several minutes. It depends how often the router sends RAs.

> Otherwise, I have minor comments. And I would also like to see a basic test application (maybe #13691 will help ?)

Yes, it will come at latest tomorrow :wink: I'm actually trying IEEE802.1x/EAP WPA2 Enterprise mode

gschorcht commented:
I have pushed the following changes with the last push:

- The additional TX and RX buffers with a length of `ETHERNET_MAX_LEN` were removed which saves about 3 kByte of RAM (commit https://github.com/RIOT-OS/RIOT/pull/13754/commits/f8d804276dcd0ff81e0ca4d8411156813fb7c17e)
- Board definition for Feather M0 was extended (commit https://github.com/RIOT-OS/RIOT/pull/13754/commits/f86212d46a7bca826fdb2a3cbfc40cefe968b973) using a pseudomodule @aabadie could you take a look whether this would be an acceptable solution? Unfortunately, we don't have something like `PSEUDOBOARDS`.
- The WPA2 Enterprise mode was added but it seems that it is working only with TKIP as encryption algorithm instead of AES128-CCM. I have to investigate it further.

The PR is still labeled as `WIP`. So may I squash it to make the changes better readable?

gschorcht commented:
After 2 days of trying to get the WPA2 Enterprise mode with IEEE802.1X/EAP authentication working, I gave it up for the moment for the following reasons:

- At least ATWINC15x0 firmware version 19.6.1 is required but most available modules still have a firmware version below 19.5.0.
- Older firmware versions support EAP-TTLSv0/MSCHAPv2 but use an invalid SSL/TLS version (SSLv3).
- At least version 19.6.1 of the ATWINC15x0 vendor driver is required. However, the driver version used by this `netdev` driver is version 19.5.2, which is the last version available as open source. 

Even though I tried to implement the missing functions in driver version 19.5.2, I wasn't able to connect to an `hostapd` AP with EAP-TTLSv0/MSCHAPv2. BTW, I had the same problem with a module with firmware version 19.6.1 and driver version 19.6.1. Unfortunately, there are no detailed information what the problem is at the ATWINC15x0, I could only see `hostapd` log messages and the WiFi packets in wireshark.

gschorcht commented:
Since WPA2 Enterprise mode can't realized, I finished the work on the driver for the moment. So it in a state where it could be reviewed. I have squashed and rebased before the review is started to make it better readable.

@benpicco Your comments were addressed before the rebase. I added `NETOPT_RSSI` so that it is shown in `ifconfig` comman. I will think about an interface for WiFi `netdevs` to scan for APs.

@aabadie Please take a look on the approach how the WiFi variant of Feather M0 is addressed with the existing board definition.

benpicco commented:
@aabadie want to give this another look?

aabadie commented:
ping @gschorcht, if you have some time to rebase and address the remaining comments, I think we are almost good here !

gschorcht commented:
> ping @gschorcht, if you have some time to rebase and address the remaining comments, I think we are almost good here !

@aabadie After having spent the last 3 months completely preparing my courses for the ad-hoc online summer term at the university, I am slowly, very slowly resuming my work for RIOT. I will try to address your comments in the next days.

aabadie commented on boards/feather-m0/include/board.h:
```diff
@@ -42,6 +42,20 @@ extern "C" {
 #define LED0_TOGGLE         (LED_PORT.OUTTGL.reg = LED0_MASK)
 /** @} */
 
+#ifdef MODULE_FEATHER_M0_WIFI
```
Wrapping the define in an ifdef block is not required so you can remove this.

aabadie commented on pkg/driver_atwinc15x0/patches/0002-remove-platform-files.patch:
```diff
@@ -0,0 +1,885 @@
+From 6c56be47ce245340129438daafa09b4b0a45b2d2 Mon Sep 17 00:00:00 2001
```
ok

aabadie commented on pkg/driver_atwinc15x0/Makefile:
```diff
@@ -0,0 +1,17 @@
+include $(RIOTBASE)/pkg/pkg.mk
+
+all:
+	"$(MAKE)" MODULE=driver_atwinc15x0 \
+	          -C $(PKG_BUILDDIR)/src/driver/source \
+	          -f $(CURDIR)/Makefile.driver_atwinc15x0
+	"$(MAKE)" MODULE=driver_atwinc15x0_common \
+	          -C $(PKG_BUILDDIR)/src/common/source \
+	          -f $(CURDIR)/Makefile.driver_atwinc15x0
+	"$(MAKE)" MODULE=driver_atwinc15x0_spi_flash \
+	          -C $(PKG_BUILDDIR)/src/spi_flash/source \
+	          -f $(CURDIR)/Makefile.driver_atwinc15x0
```
Sorry to bother you with this. With the following change the build output will be a bit nicer (all `make -C xxx` are aligned).

```suggestion
	"$(MAKE)"  -C $(PKG_BUILDDIR)/src/driver/source -f $(CURDIR)/Makefile.driver_atwinc15x0 MODULE=driver_atwinc15x0
	"$(MAKE)" -C $(PKG_BUILDDIR)/src/common/source -f $(CURDIR)/Makefile.driver_atwinc15x0 MODULE=driver_atwinc15x0_common
	"$(MAKE)" -C $(PKG_BUILDDIR)/src/spi_flash/source -f $(CURDIR)/Makefile.driver_atwinc15x0 MODULE=driver_atwinc15x0_spi_flash
```

gschorcht replied:
I made these line breaks by intention because of the maximum length of line of 100 characters.

gschorcht replied:
Indeed

aabadie commented:
> How can I make sure, if at all possible, that the driver's include path is defined before the CPU's include paths when compiling the driver, so that the driver package includes the correct header?

That sounds difficult. The INCLUDES for the current board/cpu is performed globally here:

https://github.com/RIOT-OS/RIOT/blob/c0d171a0a2c5803017ef193388c5492bf3502019/Makefile.include#L366-L370

And for drivers and pkgs, this happens way after:

https://github.com/RIOT-OS/RIOT/blob/c0d171a0a2c5803017ef193388c5492bf3502019/Makefile.include#L452-L455

Changing that order would affect too many things and I don't think that doable. I would then add a patch to that renames the include in the package.

gschorcht commented:
> And for drivers and pkgs, this happens way after:
> 
> https://github.com/RIOT-OS/RIOT/blob/c0d171a0a2c5803017ef193388c5492bf3502019/Makefile.include#L452-L455
> 
> Changing that order would affect too many things and I don't think that doable.

That's clear.

But I have found another trick by adding the include path to `CFLAGS` in `pkg/atwin15x0/Makefile`.
```
CFLAGS += -I$(PKG_BUILDDIR)/src
```
With this trick the driver's package include path is the first one and the compilation of the driver package works.

I know, it is tricky and breaks the usual handling of include paths, but it helps. What do you think?

aabadie commented:
> What do you think?

Since this is only done for the package itself it shouldn't have too much impact. So I think that's ok.

gschorcht commented:
@aabadie I had to change the dependencies and the include order of the driver header files a bit to make LLVM happy. The reason was a forward type declaration that requires C11.

I'm a bit surprised that the compilation worked with `-std=c99` for all platforms but not for LLVM.
I'm wondering why we still use `-std=c99` by default even though all our platform toolchains support C11? Wasn't there a PR which switches to C11 by default (@maribu)?

aabadie commented:
> all our platform toolchains support C11?

I don't think thats the case for AVR.

gschorcht commented:
@aabadie Compilation succeed now. May I squash?

aabadie commented:
> May I squash?

Yes please !

gschorcht commented:
@aabadie Should I add proactively a `driver/atwinc15x0/Makefile.include` regarding PR #14369? It shouldn't have any effect before PR #14369 is merged but it will avoid conflicts if this PR is merged befor PR #14369.

aabadie commented:
> Should I add proactively a driver/atwinc15x0/Makefile.include regarding PR #14369?

Don't worry about #14369, I can rebase it later :)

aabadie commented on pkg/driver_atwinc15x0/Makefile:
```diff
@@ -0,0 +1,20 @@
+CFLAGS += -I$(PKG_BUILDDIR)/src
+
+all:
+	"$(MAKE)" MODULE=driver_atwinc15x0 -C $(PKG_BUILDDIR)/src/driver/source -f $(RIOTBASE)/Makefile.base
+	"$(MAKE)" MODULE=driver_atwinc15x0_common -C $(PKG_BUILDDIR)/src/common/source -f $(RIOTBASE)/Makefile.base
+	"$(MAKE)" MODULE=driver_atwinc15x0_spi_flash -C $(PKG_BUILDDIR)/src/spi_flash/source -f $(RIOTBASE)/Makefile.base
```
#14289 was merged so this Makefile must be adapted once again:
```suggestion
	"$(MAKE)" -C $(PKG_SOURCE_DIR)/src/driver/source -f $(RIOTBASE)/Makefile.base MODULE=driver_atwinc15x0
	"$(MAKE)" -C $(PKG_SOURCE_DIR)/src/common/source -f $(RIOTBASE)/Makefile.base MODULE=driver_atwinc15x0_common
	"$(MAKE)" -C $(PKG_SOURCE_DIR)/src/spi_flash/source -f $(RIOTBASE)/Makefile.base MODULE=driver_atwinc15x0_spi_flash
```
(I also propose to move the `MODULE=xxx` at the end of the line, it's purely a cosmetic change regarding the build output)

gschorcht replied:
OK

gschorcht replied:
May I squash these changes directly to avoid an extra loop in the long, long CI compilation queue?

aabadie replied:
> May I squash these changes directly to avoid an extra loop in the long, long CI compilation queue?

Sure, let's save some CI time.

gschorcht commented:
@aabadie Thanks for you patience, the reviewing, testing and merging.

Now that we have more than one WiFi network dev, we should pick up the ball again to expose the WiFi configuration via Kconfig, as started in #13135 (@leandrolanzieri).
